src/persist-txn/src/txn_read.rs

```diff
@@ -362,14 +448,167 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64, C: TxnsCodec>
+            .clone();
+        while data_upper <= self.as_of {
+            // It would be very bad if we accidentally filled any times <=
+            // latest_write with empty updates, so defensively assert on each
+            // iteration through the loop.
+            assert!(self.latest_write < data_upper);
+            debug!(
+                "CaA data snapshot {:.9} [{:?},{:?})",
+                self.data_id.to_string(),
+                data_upper,
+                self.empty_to,
+            );
+            assert!(self.latest_write <= self.as_of);
+            assert!(self.as_of < self.empty_to);
+            let res = data_write
+                .compare_and_append_batch(
+                    &mut [],
+                    Antichain::from_elem(data_upper.clone()),
+                    Antichain::from_elem(self.empty_to.clone()),
+                )
+                .await
+                .expect("usage was valid");
+            match res {
+                Ok(()) => {
+                    debug!(
+                        "CaA data snapshot {:.9} [{:?},{:?}) success",
+                        self.data_id.to_string(),
+                        data_upper,
+                        self.empty_to,
+                    );
+                    // Persist registers writes on the first write, so politely
+                    // expire the writer we just created, but (as a performance
+                    // optimization) only if we actually wrote something.
+                    data_write.expire().await;
+                    break;
+                }
+                Err(UpperMismatch { current, .. }) => {
+                    let current = current
+                        .into_option()
+                        .expect("txns shard should not be closed");
+                    debug!(
+                        "CaA data snapshot {:.9} [{:?},{:?}) mismatch actual={:?}",
+                        self.data_id.to_string(),
+                        data_upper,
+                        self.empty_to,
+                        current,
+                    );
+                    data_upper = current;
+                    continue;
+                }
+            }
+        }
+    }
+
+    /// See [ReadHandle::snapshot_and_fetch].
+    pub async fn snapshot_and_fetch<K, V, D>(
+        &self,
+        data_read: &mut ReadHandle<K, V, T, D>,
+        // TODO(txn): It's quite surprising to require a WriteHandle for reads,
+        // see what we can do about making this nicer.
+        data_write: WriteHandle<K, V, T, D>,
+    ) -> Result<Vec<((Result<K, String>, Result<V, String>), T, D)>, Since<T>>
+    where
+        K: Debug + Codec + Ord,
+        V: Debug + Codec + Ord,
+        D: Semigroup + Codec64 + Send + Sync,
+    {
+        self.unblock_read(data_write).await;
+        data_read
+            .snapshot_and_fetch(Antichain::from_elem(self.as_of.clone()))
+            .await
+    }
+
+    fn validate(&self) -> Result<(), String> {
```

Comment:
Uncovered according to coverage report.

Reply:
I checked and this looks like it should be used in a number of unit tests, is it not? That'd be very concerning if it's not

Reply:
Feel free to ignore it in this case. The coverage report is unfortunately not always 100% accurate (for example, if some tests fail their coverage won't be included and it is rather common that some tests fail in coverage mode because they run way slower in this mode).